src/DynamoCore/DSEngine/NodeToCode.cs

```diff
@@ -1,16 +1,692 @@
+        /// <param name="renamingMap"></param>
+        private static void GetInputOutputMap(
+            IEnumerable<NodeModel> nodes, 
+            out Dictionary<string, string> inputMap,
+            out Dictionary<string, string> outputMap,
+            out Dictionary<string, string> renamingMap)
+        {
+            outputMap = new Dictionary<string, string>();
+            renamingMap = new Dictionary<string, string>();
+            var inputVariableSet = new HashSet<string>();
+
+            foreach (var node in nodes)
+            {
+                // Collects all inputs of selected nodes
+                foreach (var inport in node.InPorts)
+                {
+                    if (inport.Connectors.Count == 0)
+                        continue;
+
+                    var startPort = inport.Connectors[0].Start;
+                    var inputNode = startPort.Owner;
+                    var portIndex = startPort.Index; 
+                    var cbn = inputNode as CodeBlockNodeModel;
+
+                    if (nodes.Contains(inputNode))
+                    {
+                        // internal node
+                        if (cbn != null)
+                        {
+                            string inputVar = cbn.GetAstIdentifierForOutputIndex(portIndex).Value;
+                            string originalVar = cbn.GetRawAstIdentifierForOutputIndex(portIndex).Value;
+
+                            renamingMap[inputVar] = originalVar;
+                            var key = String.Format("{0}%{1}", originalVar, cbn.GUID);
+                            renamingMap[key] = inputVar;
+                        }
+                    }
+                    else // extern node, so they should be added to inputMap
+                    {
+                        string inputVar = inputNode.GetAstIdentifierForOutputIndex(portIndex).Value;
+                        inputVariableSet.Add(inputVar);
+                    }
+                }
+
+                // Collect all outputs of selected nodes
+                var thisCBN = node as CodeBlockNodeModel;
+                if (thisCBN != null)
+                {
+                    for (int i = 0; i < thisCBN.OutPorts.Count(); ++i)
+                    {
+                        var inputVar = thisCBN.GetAstIdentifierForOutputIndex(i).Value;
+                        var originalVar = thisCBN.GetRawAstIdentifierForOutputIndex(i).Value;
+
+                        outputMap[inputVar] = originalVar;
+                        var key = String.Format("{0}%{1}", originalVar, thisCBN.GUID);
+                        outputMap[key] = inputVar;
+                    }
+                }
+                else
+                {
+                    for (int i = 0; i < node.OutPorts.Count(); ++i)
+                    {
+                        var inputVar = node.GetAstIdentifierForOutputIndex(i).Value;
+                        outputMap[inputVar] = string.Empty;
+                    }
+                }
```

Comment:
Seems like the above two `OutPorts.Count()` can be made `OutPorts.Count` since that is a valid property on `ObservableCollection` (so that we don't cause any enumeration repeatedly).